source/common/statistic_impl.h

```diff
@@ -150,4 +151,118 @@ class HdrStatistic : public StatisticImpl {
   struct hdr_histogram* histogram_;
 };
 
-} // namespace Nighthawk
\ No newline at end of file
+/**
+ * CircllhistStatistic uses Circllhist under the hood to compute statistics.
+ * Circllhist is used in the implementation of Envoy Histograms, compared to HdrHistogram it trades
+ * precision for fast performance in merge and insertion. For more info, please see
+ * https://github.com/circonus-labs/libcircllhist
+ */
+class CircllhistStatistic : public StatisticImpl {
+public:
+  CircllhistStatistic() {
+    histogram_ = hist_alloc();
+    ASSERT(histogram_ != nullptr);
+  }
+  ~CircllhistStatistic() override { hist_free(histogram_); }
+
+  void addValue(uint64_t value) override {
+    hist_insert_intscale(histogram_, value, 0, 1);
+    StatisticImpl::addValue(value);
+  }
+  double mean() const override { return hist_approx_mean(histogram_); }
+  double pvariance() const override { return pstdev() * pstdev(); }
+  double pstdev() const override {
+    return count() == 0 ? std::nan("") : hist_approx_stddev(histogram_);
+  }
+  StatisticPtr combine(const Statistic& statistic) const override;
+  uint64_t significantDigits() const override { return 1; }
+  StatisticPtr createNewInstanceOfSameType() const override {
+    return std::make_unique<CircllhistStatistic>();
+  }
+  nighthawk::client::Statistic toProto(SerializationDomain domain) const override;
+
+private:
+  histogram_t* histogram_;
+};
+
+/**
+ * In order to be able to flush a histogram value to downstream Envoy stats Sinks, abstract class
+ * SinkableStatistic takes the Scope reference in the constructor and wraps the
+ * Envoy::Stats::HistogramHelper interface. Implementation of sinkable Nighthawk Statistic class
+ * will inherit from this class.
+ */
+class SinkableStatistic : public Envoy::Stats::HistogramImplHelper {
+public:
+  // Calling HistogramImplHelper(SymbolTable& symbol_table) constructor to construct an empty
+  // MetricImpl. This is to bypass the complicated logic of setting up SymbolTable/StatName in
+  // Envoy.
+  SinkableStatistic(Envoy::Stats::Scope& scope, const absl::optional<int> worker_id)
+      : Envoy::Stats::HistogramImplHelper(scope.symbolTable()), scope_(scope),
+        worker_id_(worker_id) {}
+
+  ~SinkableStatistic() override {
+    // We must explicitly free the StatName here in order to supply the
+    // SymbolTable reference.
+    MetricImpl::clear(scope_.symbolTable());
+  }
+
+  // Currently Envoy Histogram Unit supports {Unspecified, Bytes, Microseconds, Milliseconds}. By
+  // default, Nighthawk::Statistic uses nanosecond as the unit of latency histograms, so Unspecified
+  // is returned here to isolate Nighthawk Statistic from Envoy Histogram Unit.
+  Envoy::Stats::Histogram::Unit unit() const override {
+    return Envoy::Stats::Histogram::Unit::Unspecified;
+  };
+  Envoy::Stats::SymbolTable& symbolTable() override { return scope_.symbolTable(); }
+
+  const absl::optional<int> worker_id() { return worker_id_; }
```

Comment:
Since this one isn't an override and is public, we should document it. We should also explicitly say why does it return absl::optional and what does it mean when the value isn't present.

Reply:
Done. PTAL